### PR TITLE
Key the RBAC audit log's hash chain with the server secret

### LINKED ISSUE #472: Audit log: key the hash chain with the server secret and report tail-check read errors
Follow-ups from the security review of PR #470 (#65).

1. The `audit_events` hash chain is a plain SHA-256 over public fields, so anyone who can write `auth.db` can rewrite rows and recompute every later hash. Keying it with HMAC-SHA256 using the existing server secret file would raise the bar to "can write that file and read the secret". This changes the stored hash values, so it needs a schema bump or a re-chain step.
2. `verifyAuditTail` swallows any error from the `sqlite_sequence` query rather than only the no-row case, so a genuine read failure silently skips the tail check; it should be reported.
3. Optional: emit a `[AUDIT]` server log line on each successful `GET /api/v1/rbac/audit` naming the actor, filter and row count, since successful reads are deliberately not recorded in the table.

### PR DESCRIPTION
**Pull request by dpage**
## Summary

The `audit_events` hash chain was a plain SHA-256 over public fields, so anyone able to write `auth.db` could rewrite rows and recompute every later hash. Events are now hashed with HMAC-SHA256 under a key derived from the server secret, so an undetected rewrite needs read access to that file as well.

- The hash version is bound into the digest and may never fall as the chain advances, so an event cannot be relabelled into an encoding the verifier will not check. Version 1 is refused outright rather than verified, because anyone can compute an unkeyed hash, which makes an inherited log and a forged one indistinguishable.
- An upgraded installation re-hashes its existing events once, under the operator's hand, with `-rechain-audit-log`; the server will not open a database holding unkeyed events until it has. A database holding keyed and unkeyed events together is refused whatever the order, since no server path produces that shape and re-chaining would sign the forgery.
- The retention purge now deletes a contiguous prefix of identifiers rather than by `occurred_at` directly. `occurred_at` is a value in the file, so backdating the newest rows would otherwise have steered the purge into removing them, relinking the chain and restoring the `sqlite_sequence` agreement the tail check reads to detect truncation.
- A successful read of `GET /api/v1/rbac/audit` is now noted in the server log with the reader, the filters and the row count, since successful reads are deliberately not recorded in the table.

## Breaking changes

**Every command-line subcommand now requires the server secret**, read-only ones included, because the store resolves the key when it opens. An installation that has never created a secret file will see them stop working at this upgrade.

**Re-chaining changes every stored hash.** A hash recorded out of band, or a verification report taken before the re-chain, cannot be reconciled against the log afterwards.

## Item 2 needed no change

The issue also asked for `verifyAuditTail` to report read errors from the `sqlite_sequence` query rather than swallowing them. `git log -S "failed to read newest audit row and sequence"` returns only bf608725, so it was fixed during PR #470's review, after this issue was filed.

## Test plan

- [x] `make test-all` in `server/`: tests, lint and the coverage ratchet all pass
- [x] Six rounds of security review; every finding either fixed here or filed as a follow-up
- [x] Each new test mutation-checked: the behaviour it names was broken, the test confirmed to fail, and the break reverted

<details>
<summary>Residual limits, stated in the admin guide</summary>

An attacker holding the server secret can forge freely. Deletion of a prefix of the log, up to and including all of it, is still accepted. The re-chain is a one-time operator-signed trust event and attests nothing about the period before it ran. Nothing verifies the chain at server startup; detection depends on an operator running `-verify-audit-log`.

</details>

Closes #472

## Summary by CodeRabbit

- **Security**
  - Audit logs now use server-secret-backed HMAC-SHA256 protection.
  - Invalid, downgraded, mixed, or tampered audit chains are rejected.
  - The server requires a valid audit secret before opening the authentication store.

- **New Features**
  - Added audit-log verification with distinct failure statuses.
  - Added a confirmed re-chaining command for eligible legacy logs.
  - Audit-log reads are now recorded with sanitized filters and result counts.
  - Retention cleanup removes only the oldest contiguous events.

- **Documentation**
  - Expanded configuration, administration, and security guidance for audit logs and re-chaining.

### CODE REVIEW

File: docs/admin-guide/managing-users-and-permissions/audit-log.md [hunk 1]
```diff
@@ -285,26 +305,42 @@ Each line of the JSON output is one complete event, including the
 `jq` or a log shipper:
 
 ```json
-{"id":2,"occurred_at":"2026-09-15T11:09:17.648856224Z","actor_type":"cli","actor_id":null,"actor_name":"dba-ops","action":"user.create","target_type":"user","target_id":1,"target_name":"jane.doe","outcome":"success","details":{"after":{"id":1,"username":"jane.doe","display_name":"","email":"","annotation":"","enabled":true,"is_superuser":false,"is_service_account":false}},"prev_hash":"76bed6cc892595f6035701e0b68b3c112088d4a14b2e2db9173043735f9f85e7","hash":"5438335f26c3c01a4f42976957dca65f25b48cf44c637ff38ac8bd042736a092","hash_version":1}
+{"id":2,"occurred_at":"2026-09-15T11:09:17.648856224Z","actor_type":"cli","actor_id":null,"actor_name":"dba-ops","action":"user.create","target_type":"user","target_id":1,"target_name":"jane.doe","outcome":"success","details":{"after":{"id":1,"username":"jane.doe","display_name":"","email":"","annotation":"","enabled":true,"is_superuser":false,"is_service_account":false}},"prev_hash":"76bed6cc892595f6035701e0b68b3c112088d4a14b2e2db9173043735f9f85e7","hash":"5438335f26c3c01a4f42976957dca65f25b48cf44c637ff38ac8bd042736a092","hash_version":2}
```
[the remark this answers is not in the document]
Comment (dpage) [reply docs/admin-guide/managing-users-and-permissions/audit-log.md]: Wrapped in 08a0c0c7, still valid JSON with every value kept. The two 64-character hashes need their values on a continuation line to fit. I also adjusted the lead-in, since "each line of the JSON output is one complete event" read oddly beside a sample now spread over 27 lines.

File: docs/admin-guide/managing-users-and-permissions/audit-log.md [hunk 2]
```diff
@@ -394,6 +545,20 @@ the event that preceded it, the chain stays verifiable across a purge;
...
+a log the server wrote itself, where identifiers and timestamps rise
+together, the two are the same set of events. They part company when an
+event's timestamp has been altered in the database, which is the point:
+identifiers are assigned by SQLite in insertion order and nothing the
+server does writes them, so a backdated event cannot steer the purge
+into removing the newest events instead of the oldest. One consequence
+is worth knowing about, since it shows up on a server that has been idle
+for longer than the retention period: when no event at all falls inside
+the window, the purge removes nothing rather than emptying the log, and
+the backlog clears as soon as the next event is recorded.
```
[the remark this answers is not in the document]
Comment (dpage) [reply docs/admin-guide/managing-users-and-permissions/audit-log.md]: Correct, and the code agrees: `startTokenCleanup` purges once at startup and then on each five-minute tick, and recording an event runs no purge. Reworded in 08a0c0c7 to say the backlog clears at the first purge to run once an event falls inside the window again. The same sentence was in the changelog and is fixed there too.

File: server/src/internal/api/rbac_audit_handlers.go [hunk 3]
```diff
@@ -65,10 +67,76 @@ func (h *RBACHandler) handleAudit(w http.ResponseWriter, r *http.Request) {
...
+	add("target_type", f.TargetType)
+	if f.TargetID != nil {
+		add("target_id", strconv.FormatInt(*f.TargetID, 10))
+	}
+	add("outcome", f.Outcome)
+	if f.Since != nil {
+		add("since", f.Since.UTC().Format(time.RFC3339))
+	}
+	if f.Until != nil {
+		add("until", f.Until.UTC().Format(time.RFC3339))
```
[the remark this answers is not in the document]
Comment (dpage) [reply server/src/internal/api/rbac_audit_handlers.go]: Fixed in 08a0c0c7. Confirmed the premise first: `parseOptionalTime` parses with the `time.RFC3339` layout, and Go's parser accepts a fractional-second field even when the layout omits one, so `08:30:00.123Z` did parse and was being logged truncated. Both formats now use `time.RFC3339Nano`.

File: server/src/internal/config/config_test.go [hunk 4]
```diff
@@ -2014,3 +2014,102 @@ func TestAuditRetentionDaysNilPointerDefaultsTo90(t *testing.T) {
...
+}
+
+// TestLoadConfigSecretFileUnreadable checks that a file that exists but
+// cannot be read is reported rather than treated as absent: falling
+// through to the default search order there would look for the secret
+// somewhere the operator did not put it.
+func TestLoadConfigSecretFileUnreadable(t *testing.T) {
+	if os.Geteuid() == 0 {
+		t.Skip("root can read a 0000 file")
+	}
```
[the remark this answers is not in the document]
Comment (dpage) [reply server/src/internal/config/config_test.go:2112]: Fixed in 08a0c0c7. Swept the other tests added on this branch for the same `Geteuid`-plus-mode-0000 shape and this was the only one.

**Comment by dpage** [comment]
Codacy: the last remaining finding (Opengrep `concat-sqli` on `ListAuditEvents` in `server/src/internal/auth/audit.go`) is marked as a false positive. The other seven were fixed at source by lifting the concatenated fragments into package-level constants; this one cannot be, because the `WHERE` clause genuinely is assembled at runtime. It is safe nonetheless: `auditWhere` builds its clauses solely from fixed literals (`actor_name = ?` and the like) joined with ` AND `, and every filter value goes in as a bound parameter, so nothing untrusted reaches the SQL text. This follows the existing precedent for the same rule elsewhere in the repo.

**Review by AntTheLimey — CHANGES_REQUESTED** [review]
Two blocking items, and they compound. `-verify-audit-log` reports a
**detected tail truncation with exit status 1**, which the table you
added to the admin guide defines as "the check could not run, for
example the store would not open" — and separately, that truncation
turns out to need no secret to commit at all, which contradicts what
the residual-limits section and changelog claim about it.

On the exit code: `describeAuditVerifyFailure` treats a failure as
tampering only when it wraps `ErrAuditChainBroken`,
`ErrAuditChainDowngraded` or `ErrAuditUnkeyedRow`, and none of
`verifyAuditTail`'s four returns wraps any of them. Driving
`VerifyAuditChain` and `describeAuditVerifyFailure` directly on this
head: deleting the newest `audit_events` row gives `exit=1, "audit
chain tail missing: newest row 1, sequence 2"`, whilst dropping the
`audit_events_no_update` trigger gives `exit=2`. A scheduled job that
treats 1 as an infrastructure retry and 2 or 3 as an incident misses the
one condition the `sqlite_sequence` comparison exists to find. The
branch is untested too: `describeAuditVerifyFailure` measures 66.7%,
and the uncovered blocks are exactly the `firstBad == 0` arm.

On the deeper issue: `sqlite_sequence` — the table the tail check reads
— has no protection anywhere in this codebase, unlike `audit_events`
itself, which has the `audit_events_no_update` trigger. So an attacker
with only write access to `auth.db` can delete the newest rows and run
one `UPDATE sqlite_sequence SET seq = <new MAX(id)> WHERE name =
'audit_events'`, and the tail check passes cleanly — no hash
recomputed, no secret touched, because nothing about a hash chain
proves what came after its last remaining link. The changelog says an
undetected rewrite "now needs read access to the secret file as well",
and the admin guide files the `sqlite_sequence` fix-up under "an
attacker holding both the server secret and write access" — neither
covers this. Combined with the exit-code bug, the one attack this PR
doesn't require a secret for is also the one with no working detection:
manual-only, defeated by an ordinary UPDATE, and misreported when it
does fire. Fixing the exit code is necessary but not sufficient — the
residual-limits wording needs to move this caveat out from under
"holding the server secret", and hardening `sqlite_sequence` itself is
worth its own follow-up issue rather than blocking this PR further.

Six more optional points, none blocking. The re-chain isn't restricted
to the upgrade case it's documented as — `auditRechainPlan` only calls
`checkNoKeyedAuditRows` when `UnkeyedEvents > 0`, so
`-rechain-audit-log -confirm-rechain` will silently re-sign an already
healthy, fully keyed log with no prompt at all; gate it on "the log
doesn't verify under the current key" instead, which also covers the
next point. A replaced or rotated secret opens the store cleanly and
starts appending under the new key with no rotation procedure and only
an easy-to-dismiss exit-3 signal; verifying the newest row's HMAC at
open (one query) would close this. The secret file itself is only
warned about, not refused, when group/world-readable, which matters
more now that it's the whole basis for tamper evidence — your sibling
`ReadOwnerOnlyFile` already hard-refuses that case elsewhere.
`AuditKeyForTesting` and `SeedUnkeyedAuditLogForTesting` ship in
`ai-dba-server` and enforce their test-only contract with `panic`,
which the cross-project standard bans; an `internal/auth/authtest`
package would keep the published key out of the binary entirely with
no panic needed. The new audit-read log line has no length cap on
caller-supplied fields, unlike the CLI's own `printAuditTable`, which
truncates at 20-24 characters. And both knowledge-base files cite
`audit.go:1589-1601`, which already points at the wrong lines on this
head; name the function instead, and the security-auditor README's own
limits list should gain the `sqlite_sequence` caveat above too.

Everything else I checked holds. The version number is inside the
digest, `auditHash` refuses version 1 outright, and `VerifyAuditChain`
refuses a version that falls, so the relabel-to-unkeyed downgrade is
genuinely closed; `ensureNoUnkeyedAuditRows` and `auditRechainPlan`
both route through `checkNoKeyedAuditRows`, which consults no ordering,
so the id -1 trick does not defeat the mixed-log refusal in either
direction; the server does refuse to start on unrechained data and
`main` exits 1 with a named error; the identifier-based purge can
express only a prefix of ids, so no amount of backdating steers it into
the newest rows; every CLI subcommand really does need the secret,
since the only two non-test `NewAuthStore` call sites are
`initAuthStore` and `openAuthStoreCLI`; `describeAuditFilter` can
render only names, types, numbers and timestamps, each quoted, with the
actor name sanitised; issue #472 item 2 is indeed already fixed on
`main` by `bf608725`; the key derivation matches this codebase's own
OIDC-state precedent (PBKDF2-HMAC-SHA256, 100,000 iterations); and the
admin guide documents the other three residual limits as stated.
`make coverage` passes at 66.5% against the raised 65.5 ratchet, with
2897 tests passing, 0 failing and 3 pre-existing skips; `gofmt -l` is
clean and `golangci-lint` reports 0 issues. The mutation-check claim
holds for the three security-relevant tests I read closely.

On merge order, land #528 first. `git merge-tree` reports three
textual conflicts, in `docs/changelog.md`, `server/Makefile` (both
raise `COVERAGE_MIN` from 64.8) and `rbac_audit_handlers.go`, all
mechanical. The real coupling is the constructor: #528 adds seven
three-argument `auth.NewAuthStore` call sites across five test files
this PR does not touch, so whichever lands second must add
`auth.AuditKeyForTesting()` to each. Nothing in this PR's core logic
depends on #528.

**Comment by dpage** [comment]
@AntTheLimey thanks, both blocking items are addressed in 7e252570 (rebased onto `main`, which also picked up the `initAuthStore` call sites #518 added).

- **Exit status:** every disagreement `verifyAuditTail` reports (tail missing, sequence below the newest row, sequence row or table removed) now wraps `ErrAuditChainBroken`, so a truncated tail exits 2; a read that fails still exits 1. `describeAuditVerifyFailure` now covers the `firstBad == 0` arm in both directions, and there is an end-to-end test that deletes the newest row and checks the sentinel.
- **Wording:** the admin guide's limits, the changelog, the `verifyAuditTail` comment and the security-auditor notes now say that truncating the tail and writing `sqlite_sequence` down to match needs write access alone, rather than filing it under the secret. Hardening it is #544.

Of the optional points, I've capped each filter value in the audit-read log line at 64 runes, replaced the line-number references in both knowledge-base files with function names, and added the `sqlite_sequence` caveat to the security-auditor limits. I've left the other four for a follow-up rather than growing this PR further, since each changes behaviour or structure: gating the re-chain on "does not verify under the current key", checking the newest row's HMAC at open to catch a rotated secret, refusing a group- or world-readable secret file, and moving the test helpers into an `authtest` package.

File: .claude/golang-expert/rbac-patterns.md [hunk 5]
```diff
@@ -398,11 +398,150 @@ Each row stores the rendering its hash was computed under in
...
+existing rows as that bug returning;
+`TestPurgeDoesNotReanchorAForgedPrefix` in `audit_rechain_test.go` is
+the regression test.
+
+The purge deletes a contiguous id-prefix and nothing else:
+
+```go
+DELETE FROM audit_events
+    WHERE id < (SELECT COALESCE(MIN(id), 0) FROM audit_events
+                WHERE occurred_at >= ?)
```
[the remark this answers is not in the document]
Comment (dpage) [reply .claude/golang-expert/rbac-patterns.md]: Fixed in eeec133f.

File: .claude/security-auditor/README.md [hunk 6]
```diff
@@ -128,14 +128,94 @@ recorded in the `audit_events` table of the SQLite auth store
...
+  provided a route to have the server sign a forgery: the last one let
+  the retention purge re-anchor the boundary onto a backdated row.
+  `PurgeAuditEvents` now deletes and records its own event, and writes
+  no hash over a row it did not create. Treat any proposal that has an
+  unattended path re-sign existing rows as that bug returning.
+- `PurgeAuditEvents` deletes a contiguous id-prefix and can express
+  nothing else: `DELETE ... WHERE id < (SELECT COALESCE(MIN(id), 0)
+  FROM audit_events WHERE occurred_at >= ?)`. `occurred_at` is
+  attacker-writable and `id` is not, so deleting by timestamp alone
+  made retention, which runs unattended every five minutes, a
```
[the remark this answers is not in the document]
Comment (dpage) [reply .claude/security-auditor/README.md]: Fixed in eeec133f.

File: server/src/internal/auth/audit.go [hunk 7]
```diff
@@ -308,17 +528,648 @@ func auditHashV1(ev *AuditEvent) string {
...
+type auditQuerier interface {
+	Query(query string, args ...any) (*sql.Rows, error)
+}
+
+// forEachAuditEvent calls fn for every audit row in id order, reading
+// them a page at a time. fn may execute further statements on q,
+// because each page's rows are fully read and closed before fn sees
+// them.
+func forEachAuditEvent(q auditQuerier, fn func(AuditEvent) error) error {
+	after := int64(0)
```
[the remark this answers is not in the document]
Comment (dpage) [reply server/src/internal/auth/audit.go]: Fixed in eeec133f.

**Review by AntTheLimey — APPROVED** [review]
**Verdict: Approve.**

Round 2 closes all three round-1 blocking findings. The exit-code misclassification is fixed at the source and verified against a live CLI run (both a truncated tail and a deleted sequence row now exit 2, as documented). Coverage on the previously-undertested branch is now 100%. The changelog, admin guide, code comment and security KB all correctly describe the residual `sqlite_sequence` gap as needing write access alone, not the server secret — and that gap is appropriately tracked separately as #544 rather than blocking this PR, since a real fix needs an anchor outside `auth.db` and is a deeper question than the keying work here.

Two more commits landed during this review: one fixes an unrelated edge case (audit rows at id ≤ 0) that was a fail-closed re-chain risk, not a security gap; the other is a test-only refactor with no behavior change. Neither changes this verdict.

Nice work closing the exit-code gap with a real CLI-level test rather than just a unit test on the sentinel, and for filing #544 with an accurate, scoped description of what's actually left.
